codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java

```diff
@@ -111,123 +112,152 @@ public ChannelFuture writeData(ChannelHandlerContext ctx, int streamId, ByteBuf
             int payloadLength = data.readableBytes() + padding + flags.getPaddingPresenceFieldLength();
             verifyPayloadLength(payloadLength);
 
-            ByteBuf out = ctx.alloc().buffer(FRAME_HEADER_LENGTH + payloadLength);
+            writeFrameHeader(ctx, payloadLength, DATA, flags, streamId, promiseAggregator.newPromise());
 
-            writeFrameHeader(out, payloadLength, DATA, flags, streamId);
-
-            writePaddingLength(padding, out);
+            writePaddingLength(ctx, padding, promiseAggregator.newPromise());
+        } catch (Throwable t) {
+            data.release();
+            return promiseAggregator.tryFailure(t);
+        }
 
+        try {
             // Write the data.
-            out.writeBytes(data, data.readerIndex(), data.readableBytes());
+            ctx.write(data, promiseAggregator.newPromise());
 
             // Write the required padding.
-            out.writeZero(padding);
-            return ctx.write(out, promise);
-        } catch (RuntimeException e) {
-            return promise.setFailure(e);
-        } finally {
-            data.release();
+            writeZero(ctx, padding, promiseAggregator.newPromise());
```

Comment:
I think this is much less of an issue after your previous suggestion is implemented https://github.com/netty/netty/pull/3385#discussion_r24030888.  Do you agree?